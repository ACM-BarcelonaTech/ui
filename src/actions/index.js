import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS, CELL_SETS_COLOR,
} from './actionType';
import { connectionPromise } from '../components/content-wrapper/ContentWrapper';


// eslint-disable-next-line func-names
const loadCellSets = (experimentId) => function (dispatch, getState) {
  if (getState().cellSets.data) {
    return Promise.resolve();
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      data: json.cellSets,
    }),
  ).catch((e) => console.log('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (newState) => function (dispatch, getState) {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  return dispatch({
    type: UPDATE_CELL_SETS,
    data: newState,
  });
};

const createCluster = (cellSetInfo, clusterName, color) => function (dispatch, getState) {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  return dispatch({
    type: CREATE_CLUSTER,
    data: newCluster,
  });
};

// eslint-disable-next-line func-names
const loadCells = (experimentId, requestBody) => function (dispatch, getState) {
  if (getState().cells.data) {
    return Promise.resolve();
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body: requestBody,
    };

    io.emit('WorkRequest', request);

    console.log('emitted!!!');

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      console.log('response! ');
      return dispatch({
        type: LOAD_CELLS,
        data: embedding,
      });
    });
  });
};

// eslint-disable-next-line func-names
const cellSetsColor = (colorData) => function (dispatch, getState) {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return Promise.resolve();
};

// eslint-disable-next-line import/prefer-default-export
export {
  loadCellSets, updateCellSets, createCluster, loadCells, cellSetsColor,
};
