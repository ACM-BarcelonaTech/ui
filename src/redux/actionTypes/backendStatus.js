const BACKEND_STATUS = 'backendStatus';

/**
 * Starts loading pipeline status from the API.
 */
const BACKEND_STATUS_LOADING = `${BACKEND_STATUS}/backendStatusLoading`;

/**
 * Loaded pipeline status from the API.
 */
const BACKEND_STATUS_LOADED = `${BACKEND_STATUS}/backendStatusLoaded`;

/**
 * Backend status update was received.
 */
const BACKEND_STATUS_UPDATED = `${BACKEND_STATUS}/backendStatusUpdated`;

/**
 * Error state when backend status could not be loaded from the API.
 */
const BACKEND_STATUS_ERROR = `${BACKEND_STATUS}/backendStatusError`;

export {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_UPDATED,
  BACKEND_STATUS_ERROR,
};
