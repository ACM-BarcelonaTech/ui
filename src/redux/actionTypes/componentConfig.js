/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;

export {
  LOAD_CONFIG, UPDATE_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING,
};
