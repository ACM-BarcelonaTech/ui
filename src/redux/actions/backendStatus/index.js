import loadBackendStatus from './loadBackendStatus';
import updateBackendStatus from './updateBackendStatus';

export {
  loadBackendStatus,
  updateBackendStatus,
};
