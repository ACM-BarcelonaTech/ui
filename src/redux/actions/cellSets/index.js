import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import resetCellSets from './resetCellSets';

import loadCellSets from './loadCellSets';
import saveCellSets from './saveCellSets';
import updateCellSetHierarchy from './updateCellSetHierarchy';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';

export {
  createCellSet,
  deleteCellSet,
  resetCellSets,
  loadCellSets,
  saveCellSets,
  updateCellSetHierarchy,
  updateCellSetSelected,
  updateCellSetProperty,
};
