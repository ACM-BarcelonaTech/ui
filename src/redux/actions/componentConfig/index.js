import loadPlotConfig from './loadPlotConfig';
import loadComponentConfig from './loadComponentConfig';
import updatePlotConfig from './updatePlotConfig';
import savePlotConfig from './savePlotConfig';
import fetchPlotDataWork from './fetchPlotDataWork';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchPlotDataWork,
};
