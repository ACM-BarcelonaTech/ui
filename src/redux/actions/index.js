/* eslint-disable no-param-reassign */
import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, PUSH_CELL_SETS, CREATE_CLUSTER, CELL_SETS_COLOR,
  UPDATE_GENE_LIST, LOAD_GENE_LIST, SELECTED_GENES, UPDATE_GENE_EXPRESSION,
  LOAD_CELLS, BUILD_HEATMAP_SPEC, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from './actionType';
import connectionPromise from '../../utils/socketConnection';

const loadCellSets = (experimentId) => (dispatch, getState) => {
  if (getState().cellSets.data) {
    return null;
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      experimentId,
      data: json.cellSets,
    }),
  ).catch((e) => console.error('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (experimentId, newState) => (dispatch, getState) => {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  dispatch({
    type: UPDATE_CELL_SETS,
    experimentId,
    data: newState,
  });

  return dispatch(pushCellSets(experimentId));
};

const pushCellSets = (experimentId) => (dispatch, getState) => {
  // If no loaded data exists for our cell sets and some event would
  // trigger a push, we do not want to overwrite our cell sets with
  // empty data.
  if (!getState()?.cellSets?.data) {
    return null;
  }

  return fetch(
    `${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(
        getState().cellSets.data,
        (k, v) => ((k === 'title') ? undefined : v),
      ),
    },
  ).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: PUSH_CELL_SETS,
      experimentId,
      data: json,
    }),
  ).catch((e) => console.error('Error when trying to push cell sets to API: ', e));
};

const createCluster = (experimentId, cellSetInfo, clusterName, color) => (dispatch) => {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  dispatch({
    type: CREATE_CLUSTER,
    experimentId,
    data: newCluster,
  });

  return dispatch(pushCellSets(experimentId));
};

const cellSetsColor = (colorData) => (dispatch, getState) => {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return null;
};

const loadCells = (experimentId, embeddingType) => (dispatch, getState) => {
  if (getState().cells.data) {
    return null;
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const body = {
      name: 'GetEmbedding',
      type: embeddingType,
    };

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      return dispatch({
        type: LOAD_CELLS,
        experimentId,
        data: embedding,
      });
    });
  });
};

const updateGeneList = (experimentId, tableState) => (dispatch, getState) => {
  if (getState().geneList?.tableState === tableState) {
    return null;
  }
  dispatch({
    type: LOAD_GENE_LIST,
  });
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();
    const { geneList } = getState();

    const orderBy = tableState?.sorter.field
      || geneList.tableState?.sorter.field
      || 'dispersions';

    const orderDirection = tableState?.sorter.order
      || geneList.tableState?.sorter.order
      || 'ascend';

    const currentPage = tableState?.pagination.current
      || geneList.tableState?.pagination.current
      || 1;

    const currentPageSize = tableState?.pagination.pageSize
      || geneList.tableState?.pagination.pageSize
      || 1;

    const body = {
      name: 'ListGenes',
      selectFields: ['gene_names', 'dispersions'],
      orderBy,
      orderDirection: (orderDirection === 'ascend') ? 'ASC' : 'DESC',
      offset: ((currentPage - 1) * currentPageSize),
      limit: currentPageSize,
    };

    if (tableState.geneNamesFilter) {
      body.geneNamesFilter = tableState.geneNamesFilter;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const data = JSON.parse(res.results[0].body);

      const { total, rows } = data;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      if (tableState && tableState.pagination) {
        tableState.pagination.total = total;
      }
      return dispatch({
        type: UPDATE_GENE_LIST,
        experimentId,
        data: {
          rows,
          tableState: tableState || geneList.tableState,
        },
      });
    });
  });
};

const loadDiffExpr = (
  experimentId, comparisonType, firstSelectedCluster, secondSelectedCluster,
) => (dispatch) => {
  dispatch({
    type: LOAD_DIFF_EXPR,
  });
  return connectionPromise().then((io) => {
    const ComparisonTypes = {
      One: 'Versus Rest',
      Two: 'Across Sets',
    };

    const requestUuid = uuidv4();

    const body = {
      name: 'DifferentialExpression',
      maxNum: 100,
      cellSet: firstSelectedCluster.key,
    };
    if (comparisonType === ComparisonTypes.One) {
      body.compareWith = 'rest';
    } else {
      body.compareWith = secondSelectedCluster.key;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      let data = {};

      try {
        data = JSON.parse(res.results[0].body);
      } catch (error) {
        console.error(error);
        data = { rows: [] };
      }

      const { rows } = data;
      const total = rows.length;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      return dispatch({
        type: UPDATE_DIFF_EXPR,
        experimentId,
        data: {
          allData: rows,
          total,
        },
      });
    });
  });
};

const updateSelectedGenes = (genes, selected) => (dispatch, getState) => {
  const { selectedGenes, geneExperessionData } = getState();
  selectedGenes.geneList = selectedGenes.geneList || {};
  let newGenesAdded = false;
  genes.forEach((gene) => {
    if (selected) {
      if (!selectedGenes.geneList[gene]) {
        // This will present a dict of newly selected genes from gene table
        // { <Gene Name>: <Is the gene experession resolved?> }
        selectedGenes.geneList[gene] = false;
        newGenesAdded = true;
      }
    } else {
      delete selectedGenes.geneList[gene];
      const foundGene = geneExperessionData.data.findIndex((g) => g.geneName === gene);
      geneExperessionData.data.splice(foundGene, 1);
    }
  });
  dispatch({
    type: SELECTED_GENES,
    data: { newGenesAdded },
  });
  dispatch({
    type: UPDATE_HEATMAP_SPEC,
    data: {
      genes: geneExperessionData.data,
      rendering: true,
    },
  });
  setTimeout(() => {
    // Running this action with a delay to allow rerendering of heatmap
    dispatch({
      type: UPDATE_HEATMAP_SPEC,
      data: { rendering: false },
    });
  }, 50);
};

const loadGeneExpression = (experimentId) => (dispatch, getState) => {
  const { geneExperessionData } = getState();
  if (geneExperessionData?.isLoading) {
    return null;
  }
  const { newGenesAdded, geneList } = getState().selectedGenes;
  const newGeneBatch = [];
  if (newGenesAdded) {
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(geneList)) {
      if (!value) {
        newGeneBatch.push(key);
        geneList[key] = true;
      }
    }
  }
  dispatch({
    type: SELECTED_GENES,
    data: {
      newGenesAdded: false,
    },
  });
  if (newGeneBatch.length) {
    dispatch({
      type: UPDATE_GENE_EXPRESSION,
      data: {
        isLoading: true,
      },
    });
    return connectionPromise().then((io) => {
      const requestUuid = uuidv4();

      const body = {
        name: 'GeneExpression',
        cellSets: 'all',
        genes: newGeneBatch,
      };

      const request = {
        uuid: requestUuid,
        socketId: io.id,
        experimentId,
        timeout: '2021-01-01T00:00:00Z',
        body,
      };

      io.emit('WorkRequest', request);

      io.on(`WorkResponse-${requestUuid}`, (res) => {
        const heatMapData = JSON.parse(res.results[0].body);
        const { data } = getState().geneExperessionData;
        if (data) {
          Array.prototype.push.apply(heatMapData.data, data);
        }
        dispatch({
          type: UPDATE_GENE_EXPRESSION,
          data: {
            heatMapData,
            isLoading: false,
          },
        });
        return dispatch({
          type: BUILD_HEATMAP_SPEC,
          data: {
            geneExperessionData: heatMapData,
          },
        });
      });
    });
  }
};

export {
  loadCellSets,
  updateCellSets,
  pushCellSets,
  createCluster,
  loadCells,
  cellSetsColor,
  updateGeneList,
  updateSelectedGenes,
  loadGeneExpression,
  loadDiffExpr,
};
