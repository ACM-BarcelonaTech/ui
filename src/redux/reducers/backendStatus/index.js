import { HYDRATE } from 'next-redux-wrapper';
import _ from 'lodash';
import initialState from './initialState';

import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_ERROR,
} from '../../actionTypes/backendStatus';
import { EXPERIMENTS_DELETED } from '../../actionTypes/experiments';
import backendStatusLoading from './backendStatusLoading';
import backendStatusLoaded from './backendStatusLoaded';
import backendStatusError from './backendStatusError';
import experimentsDelete from './experimentsDelete';

const backendStatusReducer = (state = initialState, action) => {
  switch (action.type) {
    case BACKEND_STATUS_LOADING: {
      return backendStatusLoading(state, action);
    }
    case BACKEND_STATUS_LOADED: {
      return backendStatusLoaded(state, action);
    }
    case BACKEND_STATUS_ERROR: {
      return backendStatusError(state, action);
    }
    case EXPERIMENTS_DELETED: {
      return experimentsDelete(state, action);
    }
    case HYDRATE: {
      const { backendStatus } = action.payload;

      if (backendStatus) {
        return _.merge(state, backendStatus);
      }

      return state;
    }
    default: {
      return state;
    }
  }
};

export default backendStatusReducer;
