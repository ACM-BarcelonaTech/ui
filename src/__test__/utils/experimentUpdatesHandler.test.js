import experimentUpdatesHandler, { updateTypes } from 'utils/experimentUpdatesHandler';
import fake from '__test__/test-utils/constants';
import pushNotificationMessage from 'utils/pushNotificationMessage';

import { updateCellSetsClustering, loadCellSets } from 'redux/actions/cellSets';
import { updateProcessingSettingsFromQC, loadedProcessingConfig } from 'redux/actions/experimentSettings';
import { updateBackendStatus } from 'redux/actions/backendStatus';
import { updatePlotData } from 'redux/actions/componentConfig';
import endUserMessages from 'utils/endUserMessages';

jest.mock('redux/actions/cellSets/updateCellSetsClustering');
jest.mock('redux/actions/experimentSettings', () => ({
  updateProcessingSettingsFromQC: jest.fn(),
  loadedProcessingConfig: jest.fn(),
}));
jest.mock('redux/actions/backendStatus/updateBackendStatus');
jest.mock('redux/actions/componentConfig/updatePlotData');
jest.mock('redux/actions/cellSets/loadCellSets');

jest.mock('utils/pushNotificationMessage');

const mockDispatch = jest.fn();

const triggerExperimentUpdate = (update) => {
  const actionCreator = experimentUpdatesHandler(mockDispatch);
  actionCreator(fake.EXPERIMENT_ID, update);
};

const mockError = {
  error: true,
  errorCode: 'MOCK_ERROR_CODE',
  userMessage: 'mockUserMessage',
};

const mockQcUpdate = {
  type: updateTypes.QC,
  input: {
    taskName: 'mockTaskName',
    sampleUuid: 'mockSampleUuid',
  },
  output: {
    config: {
      processingConfigUpdate: {
        mockProcessingConfigUpdate: 'mockProcessingConfigUpdate',
      },
    },
    plotData: {
      mockPlotUuid: 'mockPlotData',
    },
  },
  response: {
    error: false,
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
};

const mockGem2sUpdate = {
  type: updateTypes.GEM2S,
  response: {
    error: false,
  },
  item: {
    processingConfig: {
      mockProcessingConfig: 'mockProcessingConfig',
    },
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
};

const mockWorkResponseUpdate = {
  type: updateTypes.WORK_RESPONSE,
  request: {
    body: {
      name: 'WorkResponse',
    },
  },
  response: {
    error: false,
  },
};

describe('ExperimentUpdatesHandler', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Triggers properly for backend status updates', () => {
    const mockBackendStatusUpdate = {
      status: {
        pipeline: 'mockPipelinStatus',
        gem2s: 'mockGem2sStatus',
        worker: 'mockWorkerStatus',
      },
    };
    triggerExperimentUpdate(mockBackendStatusUpdate);

    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateProcessingSettingsFromQC.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for QC updates ', () => {
    const mockUpdate = mockQcUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - update processin settings
    expect(updateProcessingSettingsFromQC).toHaveBeenCalledTimes(1);
    const updateParams = updateProcessingSettingsFromQC.mock.calls[0];
    expect(updateParams).toMatchSnapshot();

    // Dispatch 2 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(2);

    // Does not load cell sets
    expect(loadCellSets).not.toHaveBeenCalled();

    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Loads cell sets if QC pipeline completes ', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      status: {
        pipeline: {
          status: 'SUCCEEDED',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - update processin settings
    expect(updateProcessingSettingsFromQC).toHaveBeenCalledTimes(1);
    const updateParams = updateProcessingSettingsFromQC.mock.calls[0];
    expect(updateParams).toMatchSnapshot();

    // Dispatch 2 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    // Dispatch 3 - load cellsets on pipeline finish
    expect(loadCellSets).toHaveBeenCalledTimes(1);
    const loadCellSetsParams = loadCellSets.mock.calls[0];
    expect(loadCellSetsParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(3);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Throws error if there are errors in QC updates ', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      response: mockError,
    };

    triggerExperimentUpdate(mockUpdate);

    expect(mockDispatch).not.toHaveBeenCalled();
    expect(updateProcessingSettingsFromQC).not.toHaveBeenCalled();
    expect(updatePlotData).not.toHaveBeenCalled();
    expect(loadCellSets).not.toHaveBeenCalled();

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'error',
      mockError.userMessage,
    );
  });

  it('Triggers properly for GEM2S updates ', () => {
    const mockUpdate = mockGem2sUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Loaded processingConfig
    expect(loadedProcessingConfig).toHaveBeenCalledTimes(1);
    const updateParams = loadedProcessingConfig.mock.calls[0];
    expect(updateParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Throws error if there are errors in GEM2S updates ', () => {
    const mockUpdate = {
      ...mockGem2sUpdate,
      response: mockError,
    };

    triggerExperimentUpdate(mockUpdate);

    expect(mockDispatch).not.toHaveBeenCalled();
    expect(loadedProcessingConfig).not.toHaveBeenCalled();

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'error',
      mockError.userMessage,
    );
  });

  it('Triggers properly for WorkResponse - ClusterCells updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'ClusterCells',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update cell set clustering
    expect(updateCellSetsClustering).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED,
    );
  });

  it('Triggers properly for WorkResponse - GetExpressionCellSets updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'GetExpressionCellSets',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Load cell sets
    expect(loadCellSets).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_NEW_CLUSTER_CREATED,
    );
  });

  it('Throws error if there are errors in WorkResponse updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      response: mockError,
    };

    triggerExperimentUpdate(mockUpdate);

    expect(mockDispatch).not.toHaveBeenCalled();

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'error',
      mockError.userMessage,
    );
  });

  it('Throws specific error if GetExpressionCellSets returns empty cell set', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'GetExpressionCellSets',
        },
      },
      response: {
        error: true,
        errorCode: 'R_WORKER_EMPTY_CELL_SET',
        userMessage: 'Can not create empty cell set',
      },
    };

    triggerExperimentUpdate(mockUpdate);

    expect(mockDispatch).not.toHaveBeenCalled();

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'error',
      endUserMessages.EMPTY_CLUSTER_NOT_CREATED,
    );
  });
});
