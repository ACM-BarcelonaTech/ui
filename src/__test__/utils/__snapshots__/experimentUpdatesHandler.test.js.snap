// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  1`] = `
Array [
  "mockTaskName",
  Object {
    "processingConfigUpdate": Object {
      "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
    },
  },
  "mockSampleUuid",
  false,
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  2`] = `
Array [
  "mockPlotUuid",
  "mockPlotData",
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  3`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  true,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for GEM2S updates  1`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  Object {
    "mockProcessingConfig": "mockProcessingConfig",
  },
  true,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  1`] = `
Array [
  "mockTaskName",
  Object {
    "processingConfigUpdate": Object {
      "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
    },
  },
  "mockSampleUuid",
  false,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  2`] = `
Array [
  "mockPlotUuid",
  "mockPlotData",
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for backend status updates 1`] = `undefined`;
